<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Баланс-классов" data-toc-modified-id="Баланс-классов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Баланс классов</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Итоги" data-toc-modified-id="Итоги-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Итоги</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Upsampling" data-toc-modified-id="Upsampling-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Upsampling</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Downsampling" data-toc-modified-id="Downsampling-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Downsampling</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Итоги" data-toc-modified-id="Итоги-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Итоги</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Результаты" data-toc-modified-id="Результаты-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Результаты</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

Для начала импортируем все необходимые бибилиотеки. 

In [1]:
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score
from sklearn.utils import shuffle


Теперь изучим данные, с которыми нам предстоит работать

In [2]:
data = pd.read_csv('/datasets/Churn.csv')

print(data.info())
data.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB
None


RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

Каждый объект в наборе данных — это информация об одном из клиентов банка. Известно:

* RowNumber — индекс строки в данных
* CustomerId — уникальный идентификатор клиента
* Surname — фамилия
* CreditScore — кредитный рейтинг
* Geography — страна проживания
* Gender — пол
* Age — возраст
* Tenure — сколько лет человек является клиентом банка
* Balance — баланс на счёте
* NumOfProducts — количество продуктов банка, используемых клиентом
* HasCrCard — наличие кредитной карты
* IsActiveMember — активность клиента
* EstimatedSalary — предполагаемая зарплата
* Exited — факт ухода клиента

Целевым признаком является параметр Exited. В последующем мы его отделим.

В данных в столбце `Tenure` есть пропуски. Стоит заполнить эти пропуски медианным значением. 

In [3]:
data['Tenure'] = data['Tenure'].fillna(data['Tenure'].median())

Столбец `RowNumber`, `CustomerId` и `Surname` стоит убрать из обучающих данных. Они очень врядли влияют на уход клиента из банка и только усложнят обучение модели. 

Воспользуемся техникой прямого кодирования (OHE), чтобы категориальные признаки (страна и гендер) в численные, а после разделим данные на признаки и целевые признаки.

Так как тестовой выборки у нас нет, поделим данные на три выборки:

* Обучающая (60%)
* Валидационная (20%)
* Тестовая (20%)

Для этого два раза воспользуемся функцией `train_test_split`, которую заранее импортировали.

Параметр `random_state` будем задавать везде одинаковый, чтобы обеспечить повторяемость результатов

In [4]:
target = data['Exited']
features = data.drop(['Exited', 'RowNumber', 'CustomerId', 'Surname'], axis=1)

features = pd.get_dummies(features)

features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.2, random_state=12345)

features_train, features_test, target_train, target_test = train_test_split(
    features_train, target_train, test_size=0.25, random_state=12345)

print('Размер обучающей выборки:', features_train.shape[0])
print('Размер валидационной выборки:', features_test.shape[0])
print('Размер тестовой выборки:', features_valid.shape[0])


Размер обучающей выборки: 6000
Размер валидационной выборки: 2000
Размер тестовой выборки: 2000


## Исследование задачи

### Баланс классов

Разделим выборку на отрицательные и положительные объекты. Посмотрим, насколько данные сбалансированы:

In [5]:
features_zeros = features_train[target_train == 0]
features_ones = features_train[target_train == 1]
target_zeros = target_train[target_train == 0]
target_ones = target_train[target_train == 1]

print('Количество отрицательных объектов:', features_zeros.shape[0])
print('Количество положительных объектов:', features_ones.shape[0])

Количество отрицательных объектов: 4781
Количество положительных объектов: 1219


Как можно видеть, данные не сбалансированы, положительных объектов приметро в 4 раза меньше, но бороться с дисбалансом будем в третьем пункте, а тут попробуем не обращать на него внимания.

А сейчас исследуем модели трех разных типов:

* Логистическая регрессия
* Решающее дерево
* Случайный лес

Для первых двух типов попробуем найти лучшие гиперпараметры.

### Логистическая регрессия

In [6]:
model = LogisticRegression(random_state=12345, solver='liblinear', max_iter=1000)
model.fit(features_train, target_train)
predicted_valid = model.predict(features_valid)

print('F1-мера:', f1_score(predicted_valid, target_valid))

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print('AUC-ROC:', roc_auc_score(target_valid, probabilities_one_valid))

F1-мера: 0.09896907216494845
AUC-ROC: 0.6712929395492733


### Решающее дерево

In [7]:
best_model = None
best_result = 0
best_depth = 0

for depth in range(1,10):
    
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    
    result = f1_score(predicted_valid, target_valid)
    
    if result > best_result:
        best_model = model
        best_depth = depth
        best_result = result
        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        best_auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print("F1-мера наилучшей модели на валидационной выборке:", best_result, ', при глубине:', best_depth)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print('AUC-ROC:', best_auc_roc)

F1-мера наилучшей модели на валидационной выборке: 0.5608011444921316 , при глубине: 5
AUC-ROC: 0.8468610376210971


### Случайный лес

In [8]:
best_model = None
best_result = 0
best_est = 0

for est in range(1,80):
    
    model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = 20)
    model.fit(features_train, target_train)
    predicted_valid = model.predict(features_valid)
    
    result = f1_score(predicted_valid, target_valid)
    
    if result > best_result:
        best_model = model
        best_est = est
        best_result = result
        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        best_auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print("F1-мера наилучшей модели на валидационной выборке:", best_result, ', n_estimators:', best_est)
print('AUC-ROC:', best_auc_roc)

F1-мера наилучшей модели на валидационной выборке: 0.5751445086705202 , n_estimators: 61
AUC-ROC: 0.850306921096787


### Итоги

В данных присутствует дисбаланс, с которым нужно бороться (примерно 1 часть положительных обектов к 4 частям отрицательных объектов) . Результаты обучения моделей без учета дисбаланса представлены ниже:

In [9]:
display(pd.DataFrame({'DISBALANCE': ['', '', ''],
        'Тип модели': ['Логистическая регрессия', 'Решающее дерево', 'Случайный лес'],
        'F1-мера': [0.10, 0.56, 0.57],
        'AUC-ROC': [0.67, 0.85, 0.85]
        }))

DISBALANCE               Тип модели  F1-мера  AUC-ROC
0             Логистическая регрессия     0.10     0.67
1                     Решающее дерево     0.56     0.85
2                       Случайный лес     0.57     0.85

Не смотря на то, что достигнуть необходимого уровня F1-меры (0.59) не получилось, можно сделать вывод, что наилучший результат был получен с использованием случайного леса. 

Также стоит заметить, что показания метрики AUC-ROC соответствуют показаниям F1 метрики (Там, где больше F1, AUC-ROC тоже больше). 

В следующем пункте мы уберем дисбаланс разными способами и снова изучим результаты обучения моделей.

## Борьба с дисбалансом

### Upsampling

Попробуем победить дисбаланс, воспользовавшись техникой upsampling (увеличим количество положительных объектов до уровня отрицательных объектов. 

Для этого напишем функцию `upsample`

In [10]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    
    features_upsampled, target_upsampled = shuffle(
        features_upsampled, target_upsampled, random_state=12345)
    
    return features_upsampled, target_upsampled

Теперь воспользуемся новой функцией, чтобы получить сбалансированную выборку

In [11]:
features_upsampled, target_upsampled = upsample(features_train, target_train, 4)

features_zeros = features_upsampled[target_upsampled == 0]
features_ones = features_upsampled[target_upsampled == 1]

print('Количество отрицательных объектов:', features_zeros.shape[0])
print('Количество положительных объектов:', features_ones.shape[0])

Количество отрицательных объектов: 4781
Количество положительных объектов: 4876


Новая выборка сбалансирована. Попробуем снова обучить модели на ней

#### Решающее дерево

In [22]:
best_model_dt_u = None
best_result = 0
best_depth = 0

for depth in range(7,15):
    
    model_dt_u = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model_dt_u.fit(features_upsampled, target_upsampled)
    predicted_valid = model_dt_u.predict(features_valid)
    
    result = f1_score(predicted_valid, target_valid)
    
    if result > best_result:
        best_model_dt_u = model_dt_u
        best_depth = depth
        best_result = result
        probabilities_valid = model_dt_u.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        best_auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print("F1-мера наилучшей модели на валидационной выборке:", best_result, ', при глубине:', best_depth)

probabilities_valid = model_dt_u.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print('AUC-ROC:', best_auc_roc)

F1-мера наилучшей модели на валидационной выборке: 0.5915221579961464 , при глубине: 7
AUC-ROC: 0.8195895014076833


#### Случайный лес

In [14]:
best_model_rf_u = None
best_result = 0
best_est = 0

for est in range(1,80):
    
    model_rf_u = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = 20)
    model_rf_u.fit(features_upsampled, target_upsampled)
    predicted_valid = model_rf_u.predict(features_valid)
    
    result = f1_score(predicted_valid, target_valid)
    
    if result > best_result:
        best_model_rf_u = model_rf_u
        best_est = est
        best_result = result
        probabilities_valid = model_rf_u.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        best_auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print("F1-мера наилучшей модели на валидационной выборке:", best_result, ', n_estimators:', best_est)
print('AUC-ROC:', best_auc_roc)

F1-мера наилучшей модели на валидационной выборке: 0.6118836915297091 , n_estimators: 69
AUC-ROC: 0.8517272593278555


### Downsampling

Теперь попробуем воспользоваться техникой downsampling (уменьшим количество отрицательных объектов до уровня положительных. 

Напишем функцию `downsample`

In [15]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=12345)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=12345)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(
        features_downsampled, target_downsampled, random_state=12345)
    
    return features_downsampled, target_downsampled

Снова воспользуемся новой функцией, чтобы получить сбалансированную выборку

In [16]:
features_downsampled, target_downsampled = downsample(features_train, target_train, 0.25)

features_zeros = features_downsampled[target_downsampled == 0]
features_ones = features_downsampled[target_downsampled == 1]

print('Количество отрицательных объектов:', features_zeros.shape[0])
print('Количество положительных объектов:', features_ones.shape[0])

Количество отрицательных объектов: 1195
Количество положительных объектов: 1219


Выборка снова сбалансирована, но другим способом. Теперь обучаем модели

#### Логистическая регрессия

In [17]:
model_lr_d = LogisticRegression(random_state=12345, solver='liblinear', max_iter=1000)
model_lr_d.fit(features_downsampled, target_downsampled)
predicted_valid = model_lr_d.predict(features_valid)

print('F1-мера:', f1_score(predicted_valid, target_valid))

probabilities_valid = model_lr_d.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print('AUC-ROC:', roc_auc_score(target_valid, probabilities_one_valid))

F1-мера: 0.44630616907844634
AUC-ROC: 0.7217715220695847


#### Решающее дерево

In [18]:
best_model_dt_d = None
best_result = 0
best_depth = 0

for depth in range(1,10):
    
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_downsampled, target_downsampled)
    predicted_valid = model.predict(features_valid)
    
    result = f1_score(predicted_valid, target_valid)
    
    if result > best_result:
        best_model = model
        best_depth = depth
        best_result = result
        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        best_auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print("F1-мера наилучшей модели на валидационной выборке:", best_result, ', при глубине:', best_depth)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

print('AUC-ROC:', best_auc_roc)

F1-мера наилучшей модели на валидационной выборке: 0.5798922800718134 , при глубине: 7
AUC-ROC: 0.8202937152266511


#### Случайный лес

In [19]:
best_model_rf_d = None
best_result = 0
best_est = 0

for est in range(1,80):
    
    model_rf_d = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth = 20)
    model_rf_d.fit(features_downsampled, target_downsampled)
    predicted_valid = model_rf_d.predict(features_valid)
    
    result = f1_score(predicted_valid, target_valid)
    
    if result > best_result:
        best_model_rf_d = model_rf_d
        best_est = est
        best_result = result
        probabilities_valid = model_rf_d.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]
        best_auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
    
print("F1-мера наилучшей модели на валидационной выборке:", best_result, ', n_estimators:', best_est)
print('AUC-ROC:', best_auc_roc)

F1-мера наилучшей модели на валидационной выборке: 0.595667870036101 , n_estimators: 62
AUC-ROC: 0.8454585652797276


### Итоги

Дисбаланс был убран двумя разными способами (upsampling и downsampling). Результаты обучения моделей по сбалансированным данным представлены ниже:

In [20]:
display(pd.DataFrame({'UPSAMPLING': ['', '', ''],
        'Тип модели': ['Логистическая регрессия', 'Решающее дерево', 'Случайный лес'],
        'F1-мера': [0.45, 0.60, 0.61],
        'AUC-ROC': [0.72, 0.83, 0.85]
        }))
display(pd.DataFrame({'DOWNSAMPLING': ['', '', ''],
        'Тип модели': ['Логистическая регрессия', 'Решающее дерево', 'Случайный лес'],
        'F1-мера': [0.45, 0.58, 0.60],
        'AUC-ROC': [0.72, 0.82, 0.85]
        }))


UPSAMPLING               Тип модели  F1-мера  AUC-ROC
0             Логистическая регрессия     0.45     0.72
1                     Решающее дерево     0.60     0.83
2                       Случайный лес     0.61     0.85

DOWNSAMPLING               Тип модели  F1-мера  AUC-ROC
0               Логистическая регрессия     0.45     0.72
1                       Решающее дерево     0.58     0.82
2                         Случайный лес     0.60     0.85

Как можно видеть, наилучших результатов получилось достичь при балансировке данных методом upsampling. Случайный лес показал метрики F1 - 0.61 и AUC-ROC - 0.85, что соответствует условию получения метрики F1 больше 0.59. 

Проверим эту модель на тестовой выборке и посмотрим на результаты в следующем пункте.

## Тестирование модели

Наилучшим образом на валидационной выборке себя показала модель "Случайный лес". Теперь проведем ее тестирование на тестовой выборке. 

Сама модель у нас уже сохранена (`best_model_rf_u`). 

In [21]:
predicted_test = model_rf_u.predict(features_test)

probabilities_test = model_rf_u.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

print("F1-мера на тестовой выборке:", f1_score(predicted_test, target_test))
print('AUC-ROC:', roc_auc_score(target_test, probabilities_one_test))

F1-мера на тестовой выборке: 0.5952712100139081
AUC-ROC: 0.8413185740694527


Результаты проверки модели на тестовой выборке также соответствуют условию, что F1-мера должна быть больше 0.59.

## Результаты

В течение проекта были проведены следующие работы:

* Импортированы и подготовлены данные
* Исследован дисбаланс классов
* Обучены модели разных типов с разными гиперпараметрами без учета дисбаланса
* Данные были сбалансированы с помощью апсемплинга и даунсемплинга 
* Для каждого из способов балансировки снова обучены модели
* Из всех моделей выбрана модель с наилучшими показателями на валидационной выборке
* Проведено тестирование выбраной модели на тестовой выборке 

Наилучшей моделью оказался Случайный лес, обученный на данных, сбалансированных с помощью техники апсемплинг. Его метрики на валидационной выборке:

F1 - 0.61   
AUC-ROC - 0.85

В результате проверки лучшей модели на тестовой выборке были получены метрики:

F1 - 0.59   
AUC-ROC - 0.84

Полученные метрики соответствуют условию.

Также по проведенному исследованию можно сделать вывод, что метрика AUC-ROC следует за F1 метрикой. Там, где мы видим увеличение метрики F1, увеличивается и метрика AUC-ROC. 

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*